src/v/datalake/translation/partition_translator.h

```diff
@@ -0,0 +1,130 @@
+ * corresponding the leader that stepped down should eventually stop.
+ *
+ * The translator periodically wakes up, checks if there is new data that is
+ * pending translation and creates a reader for the data range and hands it off
+ * to low level iceberg translator that converts the data.
+ *
+ * The translator works in tandem with the datalake coordinator responsible for
+ * this ntp. A successful local iceberg translation of a data range is followed
+ * by that state checkpointing with the coordinator. A coordinator tracks the
+ * latest translated offset for a given ntp (along with other metadata) which is
+ * also then synchronized with the local translation stm that enforces
+ * max_collectible_offset across all replicas.
+ *
+ * The logic for this translator is as follows.
+ *
+ * while (!aborted && !term_changed):
+ *    sleep(interval)
+ *    reconcile_with_coordinator()
+ *    md = translate_newly_arrived_data_since_last_checkpoint()
+ *    checkpoint_with_coordinator(md)
+ *    sync_stm_with_coordinator(md)
+ */
+
+class partition_translator {
+public:
+    explicit partition_translator(
+      ss::lw_shared_ptr<cluster::partition> partition,
+      ss::sharded<coordinator::frontend>* frontend,
+      ss::sharded<features::feature_table>* features,
+      std::unique_ptr<datalake::cloud_data_io>* cloud_io,
+      std::chrono::milliseconds translation_interval,
+      ss::scheduling_group sg,
+      size_t reader_max_bytes,
+      std::unique_ptr<ssx::semaphore>* parallel_translations);
+
+    void start_translation_in_background(ss::scheduling_group);
+
+    ss::future<> stop();
+
+    std::chrono::milliseconds translation_interval() const;
+    void reset_translation_interval(std::chrono::milliseconds new_base);
+
+private:
+    bool can_continue() const;
+
+    ss::future<std::optional<kafka::offset>> reconcile_with_coordinator();
+
+    ss::future<> do_translate();
+
+    using translation_success = ss::bool_class<struct translation_success>;
+    ss::future<translation_success> do_translate_once(retry_chain_node& parent);
+    ss::future<model::record_batch_reader> make_reader();
+    ss::future<std::optional<coordinator::translated_offset_range>>
+    do_translation_for_range(
+      retry_chain_node& parent, model::record_batch_reader);
+
+    using checkpoint_result = ss::bool_class<struct checkpoint_result>;
+    ss::future<checkpoint_result> checkpoint_translated_data(
+      retry_chain_node& parent, coordinator::translated_offset_range);
+
+    kafka::offset min_offset_for_translation() const;
+    // Returns max consumable offset for translation.
+    std::optional<kafka::offset> max_offset_for_translation() const;
+
+    model::term_id _term;
+    ss::lw_shared_ptr<cluster::partition> _partition;
+    ss::shared_ptr<translation_stm> _stm;
+    ss::sharded<coordinator::frontend>* _frontend;
+    ss::sharded<features::feature_table>* _features;
+    std::unique_ptr<datalake::cloud_data_io>* _cloud_io;
```

Comment:
nit: i think it will be more natural to use a plain pointer here, we do not need an ownership and lifecycle management in here